internal/storage/minio_chunk_manager.go

```diff
@@ -17,62 +17,235 @@
+	var el errorutil.ErrorList
+	var objectsValues [][]byte
+	for _, key := range keys {
+		objectValue, err := mcm.Read(key)
+		if err != nil {
+			el = append(el, err)
+		}
+		objectsValues = append(objectsValues, objectValue)
+	}
+
+	if len(el) == 0 {
+		return objectsValues, nil
+	}
+	return objectsValues, el
+}
+
+func (mcm *MinioChunkManager) ReadWithPrefix(prefix string) ([]string, [][]byte, error) {
+	objects := mcm.Client.ListObjects(mcm.ctx, mcm.bucketName, minio.ListObjectsOptions{Prefix: prefix})
+
+	var objectsKeys []string
+	var objectsValues [][]byte
+
+	for object := range objects {
+		objectsKeys = append(objectsKeys, object.Key)
+	}
+	objectsValues, err := mcm.MultiRead(objectsKeys)
+	if err != nil {
+		log.Error(fmt.Sprintf("MinIO load with prefix error. path = %s", prefix), zap.Error(err))
+		return nil, nil, err
+	}
+
+	return objectsKeys, objectsValues, nil
+}
+
+func (mcm *MinioChunkManager) Mmap(filePath string) (io.ReaderAt, error) {
+	panic("this method has not been implemented")
 }
 
 // ReadAt reads specific position data of minio storage if exists.
-func (mcm *MinioChunkManager) ReadAt(key string, p []byte, off int64) (int, error) {
-	results, err := mcm.minio.Load(key)
+func (mcm *MinioChunkManager) ReadAt(filePath string, off int64, length int64) ([]byte, error) {
+	if off < 0 || length < 0 {
+		return nil, io.EOF
+	}
+
+	opts := minio.GetObjectOptions{}
+	err := opts.SetRange(off, off+length-1)
 	if err != nil {
-		return -1, err
+		return nil, err
 	}
 
-	if off < 0 || int64(len([]byte(results))) < off {
-		return 0, errors.New("MinioChunkManager: invalid offset")
+	object, err := mcm.Client.GetObject(mcm.ctx, mcm.bucketName, filePath, opts)
+	if err != nil {
+		return nil, err
 	}
-	n := copy(p, []byte(results)[off:])
-	if n < len(p) {
-		return n, io.EOF
+	defer object.Close()
+	return ioutil.ReadAll(object)
+}
+
+// Remove deletes an object with @key.
+func (mcm *MinioChunkManager) Remove(key string) error {
+	err := mcm.Client.RemoveObject(mcm.ctx, mcm.bucketName, key, minio.RemoveObjectOptions{})
+	return err
+}
+
+// MultiRemove deletes a objects with @keys.
+func (mcm *MinioChunkManager) MultiRemove(keys []string) error {
+	var el errorutil.ErrorList
+	for _, key := range keys {
+		err := mcm.Remove(key)
+		if err != nil {
+			el = append(el, err)
+		}
 	}
+	if len(el) == 0 {
+		return nil
+	}
+	return el
+}
+
+// RemoveWithPrefix removes all objects with the same prefix @prefix from minio.
+func (mcm *MinioChunkManager) RemoveWithPrefix(prefix string) error {
+	objectsCh := make(chan minio.ObjectInfo)
 
-	return n, nil
+	go func() {
+		defer close(objectsCh)
+
+		for object := range mcm.Client.ListObjects(mcm.ctx, mcm.bucketName, minio.ListObjectsOptions{Prefix: prefix, Recursive: true}) {
+			objectsCh <- object
+		}
+	}()
+
+	for rErr := range mcm.Client.RemoveObjects(mcm.ctx, mcm.bucketName, objectsCh, minio.RemoveObjectsOptions{GovernanceBypass: true}) {
```

Comment:
Our minio version does not support object locking, and simply ignore the corresponding http header, so this code is untested.
This could be an issue if user is using S3 without object locking enabled.